docs/rustdoc/prehashing.md

```diff
@@ -0,0 +1,183 @@
+>   - The prover backtracks and tries $t = 2$.
+> ---
+> **Step 3:** Proof construction for $t = 2$.
+> - Depth $k = 1$:
+>   - Compute $H_1(2)$ to determine the first bin to explore.
+>     - Assume $H_1(2) = 4$, so the prover looks in bin 4.
+>   - $\mathsf{bin_4}$: $\[D, E\]$
+>     - Extend the sequence with $s_1 = D$.
+>   - Current sequence: $(t, s_1) = (2, D)$.
+> - Depth $k = 2$:
+>   - Compute $H_1(t, s_1)$ to determine the next bin to explore.
+>     - Assume $H_1(2, D) = 2$, so the prover looks in bin 2.
+>   - $\mathsf{bin_2}$: $\[B\]$
+>     - Extend the sequence with $s_2 = B$.
+>   - Current sequence: $(t, s_1, s_2) = (2, D, B)$.
+> - Depth $k = 3$:
+>   - Compute $H_1(2, s_1, s_2)$ to determine the next bin to explore.
+>     - Assume $H_1(2, D, B) = 1$, so the prover looks in bin 1.
+>   - $\mathsf{bin_1}$: $\[A, C\]$
+>     - Extend the sequence with $s_3 = C$.
+>   - Current sequence: $(t, s_1, s_3, s_3) = (2, D, B, C)$.
+> ---
+> **Step 4:** Final Validation.
+> - A tuple with size $u = 3$ is obtained: $(t, s_1, s_2, s_3) = (2, D, B, C)$.
+> - If $H_2(2, D, B, C) = 1$, the sequence is valid, and the prover outputs it as proof.
+---
+
+## Functions
+#### Proving
+---
+> $\mathsf{Prove}(S_p) \rightarrow \pi$
+> - Input:
+>   - $S_p:~~$ `prover_set`, a set of elements to be proven.
+> - Output:
+>   - $\pi:~~$ `proof`, a valid proof $(t, s_1, \ldots, s_u)$ or $\bot$.
+> ---
+> - $bins \leftarrow$ { }
+> - **for** each $~~ i \in \[1,~ n_p\]:$
+>   - $bins\[i\] \leftarrow \[~\]$
+> - **for** each $~~ s \in S_p:$
+>   - $ind \leftarrow \mathsf{H_0}(s)$
+>   - $bins\[ ind \] \leftarrow bins\[ ind \] \cup \[s\]$
+> - **for** each $~~ t \in \[1,~ d\]:$
+>   - $\pi \leftarrow \mathsf{DFS}(t, \[~\], bins)$
+>   - **if** $~~ \pi ~!= \bot:$
+>     - **return** $~~ \pi.$
+> - **return** $~~ \bot.$
+---
+#### Depth fist search
+---
+> $\mathsf{DFS}(t, slist, S_p) \rightarrow \pi$
+> - Input:
+>   - $t:~~$ `search_counter`, index of the current subtree being searched.
+>   - $slist:~~$ `element_sequence`, candidate element sequence.
+>   - $bins:~~$ `bins`, a hash-based partition of $S_p$ elements.
+> - Output:
+>   - $\pi:~~$ `proof`, a valid proof $(t, s_1, \ldots, s_u)$ or $\bot$.
+> ---
+> - **if** $~~ \mathsf{len}(slist) == u:$
+>   - **if** $~~ \mathsf{H_2}(t, slist) == 1:$
+>     - $\pi \leftarrow (t, slist)$
+>     - **return** $~~ \pi.$
+>   - **return** $~~ \bot.$
+> - $ind \leftarrow \mathsf{H_1}(t, slist)$
+> - **for** each $~~ s \in bins\[ind\]:$
+>   - $slist_{new} \leftarrow slist \cup \[s\]$
+>   - $\pi \leftarrow \mathsf{DFS}(t, slist_{new}, bins)$
+>   - **if** $~~ \pi ~!= \bot:$
+>     - **return** $~~ \pi.$
+> - **return** $~~ \bot.$
+---
+#### Verification
+---
+> $\mathsf{Verify}(\pi) \rightarrow 0/1$
+>
+> - Input:
+>   - $\pi:~~$ `proof`, a proof of the form $(t, s_1, \ldots, s_u)$.
+> - Output:
+>   - $0/1$.
+> ---
+> - **if** $~~ t ~∉ \[d\]:$
```

Comment:
If we are using notation `[d]` here, we can also use it everywhere else instead of `[1, x]`.